docs/develop/middleware.md

```diff
@@ -0,0 +1,112 @@
+---
+title: Middleware
+subtitle: Handling cross-cutting, generic functionality
+---
+
+Middleware is a way to write reusable code that runs before or after (or both)
+the handling of API requests, often across several (or all) API endpoints.
+
+It's commonly used to implement cross-cutting concerns like request logging,
+authentication, tracing, and so on. One of the benefits of Encore is that
+all of these use cases are already handled out-of-the-box, so there's no
+need to use middleware for those things.
+
+Nonetheless, there are several use cases where it can be useful to write
+reusable functionality that applies to multiple API endpoints, and middleware
+is a good solution in those cases.
+
+Encore provides built-in support for middleware by defining a function with the
+`//encore:middleware` directive. The middleware directive takes a `target`
+parameter that specifies which API endpoints it applies to.
+
+## Middleware functions
+
+A typical middleware implementation looks like this:
+
+```go
+import (
+    "encore.dev/beta/errs"
+    "encore.dev/middleware"
+)
+
+//encore:middleware global target=all
+func ValidationMiddleware(req middleware.Request, next middleware.Next) middleware.Response {
+    // If the payload has a Validate method, use it to validate the request.
+    payload := req.Data().Payload
+    if validator, ok := payload.(interface { Validate() error }); ok {
+        if err := validator.Validate(); err != nil {
+            // If the validation fails, return an InvalidArgument error.
+            err = errs.WrapCode(err, errs.InvalidArgument, "validation failed")
+            return middleware.Response{Err: err}
+        }
+    }
+    return next(req)
+}
+```
+
+Middleware forms a chain, allowing each middleware to introspect and process
+the incoming request before handing it off to the next middleware by calling the
+`next` function that's passed in as an argument. For the last middleware in the
+chain, calling `next` results in the actual API handler being called.
+
+The `req` parameter provides information about the incoming request
+(see [package docs](https://pkg.go.dev/encore.dev/middleware#Request)).
+
+The `next` function returns a [`middleware.Response`](https://pkg.go.dev/encore.dev/middleware#Response)
+object which contains the response from the API, describing whether there was an error, and on success
+the actual response payload.
+
+This enables middleware to also introspect and even
+modify the outgoing response, like this:
+
+```go
+//encore:middleware target=tag:cache
+func CachingMiddleware(req middleware.Request, next middleware.Next) middleware.Response {
```

Comment:
Does middleware support DI?

This is a great example where you'd want DI to inject a caching implementation (or use an in-memory cache for that instance)

Reply:
Yep, added docs on this